sql: correctly test Pow decimal for 0

madelynnblue opened the pull request:
Return errors instead of nil when attempting to take the power of
something undefined. Correctly handle all cases of 0^y.

Improve the decimal tests to verify input is valid.

Fixes #10008

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/10080)

<!-- Reviewable:end -->

nvanbenschoten commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> and thanks for the cleanup in decimal_test.go

---

Reviewed 4 of 4 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/util/decimal/decimal.go, line 429 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10080#-KUQxkv9wqWBQr0yzPzW:-KUQxkv9wqWBQr0yzPzX:b70v1xs) ([raw file](https://github.com/cockroachdb/cockroach/blob/da8166c3bcb5949587dd310fd36b5ddba1f193e0/pkg/util/decimal/decimal.go#L429)):*
> ```Go
> 	isInt := tmp.Cmp(new(inf.Dec).Round(tmp, 0, inf.RoundDown)) == 0
> 
> 	xs := x.Sign()
> ```

nit for both `.Sign` calls: I usually prefer switch statements when exhaustively handling mutually exclusive return cases. Feel free to disregard

---

*[pkg/util/decimal/decimal.go, line 438 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10080#-KUQwqshQhRudXzdNSil:-KUQwqsimkE7mHoeRpMQ:bjluaqk) ([raw file](https://github.com/cockroachdb/cockroach/blob/da8166c3bcb5949587dd310fd36b5ddba1f193e0/pkg/util/decimal/decimal.go#L438)):*
> ```Go
> 		}
> 		// undefined for y < 0
> 		return nil, errors.New("zero raised to a negative power is undefined")
> ```

These two errors can be constants, right? I dont think the extra stack traces give us anything.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10080#-:-KUQyDul91Q8neoPkY4j:b-28e4p0)*
<!-- Sent from Reviewable.io -->

madelynnblue committed:
sql: correctly test Pow decimal for 0

Return errors instead of nil when attempting to take the power of
something undefined. Correctly handle all cases of 0^y.

Improve the decimal tests to verify input is valid.

Fixes #10008

mjibson commented:
Review status: 2 of 4 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/util/decimal/decimal.go, line 429 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10080#-KUQxkv9wqWBQr0yzPzW:-KUT6xESejvMiI8-nZVQ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/da8166c3bcb5949587dd310fd36b5ddba1f193e0/pkg/util/decimal/decimal.go#L429)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> nit for both `.Sign` calls: I usually prefer switch statements when exhaustively handling mutually exclusive return cases. Feel free to disregard
</details>
Done.

---

*[pkg/util/decimal/decimal.go, line 438 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10080#-KUQwqshQhRudXzdNSil:-KUT6wy5eNSQApYIcDcR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/da8166c3bcb5949587dd310fd36b5ddba1f193e0/pkg/util/decimal/decimal.go#L438)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> These two errors can be constants, right? I dont think the extra stack traces give us anything.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10080)*
<!-- Sent from Reviewable.io -->
